### ISSUE TITLE
Many Builder APIs don't document what properties are required. 

### ISSUE THREAD

**Issue by tmccombs** [aws/aws-sdk-java-v2#4291]
### Describe the issue

The javadocs for many Builder classes and interfaces don't describe which settings are required to supply. If you fail to set some settings, then you can easily end up with a NullPointer exception at runtime. 

At the very least the documentation should enumerate which options must be supplied by the user, and if there are certain situations where an option might need to be required. 

Even better would be if the type system could encode that certain fields are required so that it is impossible to construct a built object that didn't supply the requried fields, although that would be more difficult to do, and wouldn't be backwards compatible.

### Links

One example of this is https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/sts/model/AssumeRoleRequest.html which doesn't document that `roleArn` and `roleSessionName` are required. `roleArn` is kind of obvious that you need it, but `roleSessionName` less so. It is very easy to miss that you need to provide that until you get a NullPointerException.

There are others, that I have run into as well, this is just the most recent one I've run into.

**Comment by debora-ito** [comment]
Can you share the stacktrace of the NullPointerException? I'd expect to get a `StsException` as a response from the AssumeRole call, but not a NPE.

For context: the Javadocs are all auto generated based on the models provided by the service teams. Some services put that type of information in their models, but most don't. I could raise your feedback to the particular services you want to see the doc improvements, but I would need a list of the APIs you are using.

> Even better would be if the type system could encode that certain fields are required so that it is impossible to construct a built object that didn't supply the requried fields, although that would be more difficult to do, and wouldn't be backwards compatible.

As a general rule, the SDKs don't perform attribute validation on the client side. If the service change the validation rule in the future, it would force everyone to upgrade to an SDK version that matches the new rule, which is not a good experience.

**Comment by tmccombs** [comment]
Oh, you are right, it was an StsException in that case. I got confused because the error message was "Value null at 'roleSessionName' failed to satisfy constraint: Member must not be null". My bad.

Although, I'm pretty sure I ran into another case where a builder did throw NullPointerException, because I failed to set something. In that case it actually threw an exception when I called `.build()`. Unfortunately, I don't remember what class that was :grimacing:
